### ISSUE TITLE
Improve performance of fill/undo-fill in large arrays

### ISSUE THREAD

**Issue by jni** [napari/napari#2379]
Currently, labels fill [relies on doing `data == label`](https://github.com/napari/napari/blob/1c5b9b018d6225be0bda8a2fa5c60d3de37741e7/napari/layers/labels/labels.py#L776) to find pixels of the same label. This is wasteful in both 2D and 3D, but in 3D it could be unnecessarily slow by orders of magnitude if we are filling e.g. 1 nucleus within a much larger image. The correct way to do it is to use a flood fill algorithm. Some short discussion of this issue can be found [here](https://github.com/napari/napari/pull/2339#issuecomment-790183358).

In #2339 we discussed whether to use the scikit-image implementation of flood fill or code our own based on numba. For now, we are leaning towards the latter.

One issue with flood fill is that it relies on the full array being instantiated. Again, this is no worse than the current approach, but in the long term, we probably want to fill across lazy array chunks — fill in the current chunk, and only load adjacent chunks if we hit the edges. Alternatively, we could use region growing — load increasingly larger regions as we flood, without bothering to align with chunks — the alignment might not matter too much if we use caching.

**Comment by psobolewskiPhD** [comment]
skimage is a dependency for sample data and looking more closely, only networks is a skimage direct dependency that is not already a napari dependency.
Is it worth revisiting this @jni ?

**Comment by jni** [comment]
Yes we can use skimage here, though we still don't have a solution re chunking

**Comment by Czaki** [comment]
The skimage is also generating additional array in process: 
https://github.com/scikit-image/scikit-image/blob/v0.24.0/skimage/morphology/_flood_fill.py#L19-L126
https://github.com/scikit-image/scikit-image/blob/a160f384523ac70173d069954b7084bee79fb68e/skimage/morphology/_flood_fill_cy.pyx#L49-L52

Also, if I good understand, the code is not as fast as it could be because of lack of variable declaration. For example, the `i` variable is not declared

https://github.com/scikit-image/scikit-image/blob/a160f384523ac70173d069954b7084bee79fb68e/skimage/morphology/_flood_fill_cy.pyx#L75

Custom cython looks best. I expect that numba will explode as lack of dask/zarr etc support.

**Comment by jni** [comment]
@Czaki why not improve the skimage flood fill as you did with map_array? That would have a lot of impact.

**Comment by Czaki** [comment]
@jni Because I have checked it for first time 5 minutes ago? So I only have time to find weak ppint of algorithms. 

I need to find time to investigate general code shape to understand assumptions. 
For example, the queue implementation is also problematic (it may consume a huge amount of memory without need)

https://github.com/scikit-image/scikit-image/blob/main/skimage/morphology/_queue_with_history.pxi

In general, I will put it on my todo-list of things to do in my free time, but it is quite a big task.

**Comment by psobolewskiPhD** [comment]
So the issue is the size of the area being filled. Filling more than 1M pixels (e.g. 1K by 1K) blocks for ~10s and goes up from there. Undo is instant for me though.

Edit: ran into this when testing https://github.com/napari/napari/issues/8590 so all 3: paint, fill, polygon have some overlapping code paths and are very slow for 1M pixels being modified.
